src/aio/native_linux_aio_provider.cpp

```diff
@@ -68,15 +70,43 @@ error_code native_linux_aio_provider::flush(dsn_handle_t fh)
 error_code native_linux_aio_provider::write(const aio_context &aio_ctx,
                                             /*out*/ uint32_t *processed_bytes)
 {
-    ssize_t ret = pwrite(static_cast<int>((ssize_t)aio_ctx.file),
-                         aio_ctx.buffer,
-                         aio_ctx.buffer_size,
-                         aio_ctx.file_offset);
-    if (ret < 0) {
-        return ERR_FILE_OPERATION_FAILED;
-    }
-    *processed_bytes = static_cast<uint32_t>(ret);
-    return ERR_OK;
+    dsn::error_code resp = ERR_OK;
+    uint32_t buffer_offset = 0;
+    do {
+        // ret is the written data size
+        uint32_t ret = pwrite(static_cast<int>((ssize_t)aio_ctx.file),
+                              (char *)aio_ctx.buffer + buffer_offset,
+                              aio_ctx.buffer_size - buffer_offset,
+                              aio_ctx.file_offset + buffer_offset);
+        if (dsn_unlikely(ret < 0)) {
+            if (errno == EINTR) {
+                dwarn_f("write failed with errno={} and will retry it.", strerror(errno));
+                continue;
+            }
+            resp = ERR_FILE_OPERATION_FAILED;
+            derror_f("write failed with errno={}, return {}.", strerror(errno), resp);
+            return resp;
+        }
+
+        // mock the `ret` to reproduce the `write incomplete` case in the first write
+        FAIL_POINT_INJECT_OFF_F("aio_pwrite_incomplete", [&]() -> void {
+            if (dsn_unlikely(buffer_offset == 0)) {
+                --ret;
```

Comment:
I'm curious about why the test would pass if `--ret` every loop?

Reply:
` --ret` only run at first time (`dsn_unlikely(buffer_offset == 0`))